overlord/snapstate/handlers.go

```diff
@@ -777,6 +785,157 @@ func (m *SnapManager) doDownloadSnap(t *state.Task, tomb *tomb.Tomb) error {
 	return nil
 }
 
+func waitForPreDownload(task *state.Task, snapsup *SnapSetup) error {
+	st := task.State()
+	st.Lock()
+	defer st.Unlock()
+
+	tasks, err := findTasksMatchingKindAndSnap(st, "pre-download-snap", snapsup.InstanceName(), snapsup.Revision())
+	if err != nil {
+		return err
+	}
+
+	// if there is a pre-download task for the same snap, wait for it to finish
+	for _, preTask := range tasks {
+		switch preTask.Status() {
+		case state.DoingStatus:
+			var taskIDs []string
+			if err := preTask.Get("waiting-tasks", &taskIDs); err != nil && !errors.Is(err, &state.NoStateError{}) {
+				return err
+			}
+
+			if !strutil.ListContains(taskIDs, task.ID()) {
+				taskIDs = append(taskIDs, task.ID())
+				preTask.Set("waiting-tasks", taskIDs)
+			}
+
+			logger.Debugf("Download task %s will wait 1min for pre-download task %s", task.ID(), preTask.ID())
+			return &state.Retry{After: time.Minute}
```

Comment:
sorry when I commented on this I forgot that the main way to unblock this is the At() in doPreDownloadSnap, so this could be 2-3 mins